### DISCUSSION TITLE
Using focusAtom with a potential null valued atom

### DISCUSSION THREAD

**Discussion by wiredprairie** [pmndrs/jotai#1464] [Q&A] [answered]
Given the sample code below, I wanted to use a `focusAtom` to access just the `blocks` from `JType1`.
 
However, there's a TypeScript error:

```
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'WritableAtom<JType1 | null, JType1, void>' is not assignable to parameter of type 'WritableAtom<JType1 | null, JType1 | null, void>'.
      Type 'JType1 | null' is not assignable to type 'JType1'.
        Type 'null' is not assignable to type 'JType1'.ts(2769)
```

I've got `strict: true` in `tsconfig.json` as was mentioned in a few discussions/issues. The `currentJ` atom works as expected. 

```typescript
// simple demo structure
type JType1 = { id: string; blocks: string[]}

// collection of JType1
const $jMap = new Map<string, JType1>();

// there may not be a currentJ set, so `null` is allowed
export const currentJId = atom<string | null>(null);
// also -- if no ID, then no `currentJ`

// currentJ is a WritableAtom<JType1 | null, JType1, void>
export const currentJ = atom<JType1 | null, JType1>(
  (get) => {
      const jId = get(currentJId);
      if (jId === null) {
          return null;
      }
      return $jMap.has(jId) ? $jMap.get(jId)! : null;
  },
  (_get, set, jt) => {
      set(currentJId, jt.id);
  }
);

const jtBlocksAtom = focusAtom(currentJ, (optic) => optic.prop("blocks"));
```

Experimentally, to confirm that `focusAtom` is working functionally and to make TypeScript ignore the error for a moment:

```typescript
focusAtom<any, any, any>(currentJ, (optic) => optic.prop("blocks"));
```

That does work as expected.

(I do realize that if the `atom` being focused is `null`/`undefined` that a runtime error will occur. In the context of the code I'm writing, the `focusAtom` is only displayed when the corresponding `atom` is set to a value.

Is there a way to accomplish use of `focusAtom` with the proper types?

**Comment by dai-shi** [comment, marked as answer]
https://tsplay.dev/w17BAm
☝️ First, you need a symmetric atom (value and update has same type). But, it doesn't solve the typing.

https://tsplay.dev/WKOapW
☝️ Here's what I've got. There might be some better ways. I'm not very familiar with optics...

**Reply by wiredprairie** [reply]
Thanks! I'll give that a try. I hadn't heard of `optics-ts` until I read about it on your documentation pages. 😄
